qiskit/pulse/library/signal.py

```diff
@@ -0,0 +1,81 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Implements a Signal."""
+
+from typing import Any, Dict, Optional, Union
+
+from qiskit.circuit.parameterexpression import ParameterExpression
+from qiskit.pulse.library.pulse import Pulse
+from qiskit.pulse.frame import Frame
+from qiskit.pulse.exceptions import PulseError
+
+
+class Signal:
```

Comment:
This seems like a rather limited definition of a `Signal`, it seems to me that a `Signal` would be a base class of which `Pulse` and `Frame` are instances of. This strikes me more like a
```python
class Mix(Signal):
     def __init__(self, signal0: Signal, signal1: Signal, ...):
               self.signal0 = signal0
               self.signal1 = signal1

class Modulate(Signal):
     def __init__(self, pulse: Pulse, frame: Frame, ...):
                self.mixture = Mix(pulse, frame)
      
     @property
     def pulse(self):
           self.mixture.pulse
     ...
```

Which may make it easier to extend in the future.

Reply:
I see the value in `Mix` and `Modulate` and other derived constructs. Can you be more specific on who you think the `Signal` base class should look like? I would however, not see `Frame` as an instance of a signal is it is just a frequency and a phase. I.e. I do not think of playing a Frame, I would rather play a Signal which could be a mixture of pulses in different frames.